src/GC/GC.jl

```diff
@@ -3,77 +3,134 @@
-freed but is instead added to a queue of objects to free later when `enable()` is called.
+!!! note
 
-Like most PythonCall functions, you must only call this from the main thread.
+    Historically this would disable the PythonCall garbage collector. This was required
+    for safety in multi-threaded code but is no longer needed, so this is now a no-op.
 """
-function disable()
-    ENABLED[] = false
-    return
-end
+disable() = nothing
 
 """
     PythonCall.GC.enable()
 
-Re-enable the PythonCall garbage collector.
+Do nothing.
 
-This frees any Python objects which were finalized while the GC was disabled, and allows
-objects finalized in the future to be freed immediately.
+!!! note
 
-Like most PythonCall functions, you must only call this from the main thread.
+    Historically this would enable the PythonCall garbage collector. This was required
+    for safety in multi-threaded code but is no longer needed, so this is now a no-op.
 """
-function enable()
-    ENABLED[] = true
-    if !isempty(QUEUE)
-        for ptr in QUEUE
-            if ptr != C.PyNULL
-                C.Py_DecRef(ptr)
-            end
+enable() = nothing
+
+"""
+    PythonCall.GC.gc()
+
+Free any Python objects waiting to be freed.
+
+These are objects that were finalized from a thread that was not holding the Python
+GIL at the time.
+
+Like most PythonCall functions, this must only be called from the main thread (i.e. the
+thread currently holding the Python GIL.)
+"""
+function gc()
+    if C.CTX.is_initialized
+        unsafe_free_queue()
+    end
+    nothing
+end
+
+function unsafe_free_queue()
+    lock(QUEUE.lock)
+    for ptr in QUEUE.items
+        if ptr != C.PyNULL
+            C.Py_DecRef(ptr)
         end
     end
-    empty!(QUEUE)
-    return
+    empty!(QUEUE.items)
+    unlock(QUEUE.lock)
+    nothing
 end
 
 function enqueue(ptr::C.PyPtr)
     if ptr != C.PyNULL && C.CTX.is_initialized
-        if ENABLED[]
+        if C.PyGILState_Check() == 1
             C.Py_DecRef(ptr)
+            if !isempty(QUEUE.items)
+                unsafe_free_queue()
+            end
         else
-            push!(QUEUE, ptr)
+            lock(QUEUE.lock)
+            push!(QUEUE.items, ptr)
+            unlock(QUEUE.lock)
         end
     end
-    return
+    nothing
 end
 
 function enqueue_all(ptrs)
-    if C.CTX.is_initialized
-        if ENABLED[]
+    if any(!=(C.PYNULL), ptrs) && C.CTX.is_initialized
+        if C.PyGILState_Check() == 1
             for ptr in ptrs
                 if ptr != C.PyNULL
                     C.Py_DecRef(ptr)
                 end
             end
+            if !isempty(QUEUE.items)
+                unsafe_free_queue()
+            end
         else
-            append!(QUEUE, ptrs)
+            lock(QUEUE.lock)
+            append!(QUEUE.items, ptrs)
+            unlock(QUEUE.lock)
         end
     end
-    return
+    nothing
+end
```

Comment:
Tangentially related, could each of these functions have a few more inline comments? Just to help me figure out the purpose of each block.

Reply:
It won't let me suggest comments on deleted lines, but my attempt:

* `any(!=(C.PYNULL), ptrs)`: is there any work to do? if the pointers are all null then nothing to do
* `C.CTX.is_initialized`: need to ensure the context is initialized (not sure what that does?)
* `C.PyGILState_Check() == 1`: does this thread hold the GIL?
* if so, eagerly decrease the reference count on each object without putting it in the queue. Then since we know we hold the GIL, for good measure also process the queue built up on other threads
* if not, then acquire the queue lock so we can add these items to the queue